src/goto-harness/recursive_initialization.cpp

```diff
@@ -426,49 +332,421 @@ std::string recursive_initialization_configt::to_string() const
+    "__goto_harness",
+    symbol_name,
+    source_locationt{},
+    initialization_config.mode,
+    goto_model.symbol_table);
+  fresh_symbol.is_lvalue = true;
+  fresh_symbol.value = from_integer(0, signed_int_type());
+  return fresh_symbol.symbol_expr();
+}
+
+symbol_exprt recursive_initializationt::get_fresh_local_typed_symexpr(
+  const std::string &symbol_name,
+  const typet &type,
+  const exprt &init_value) const
+{
+  symbolt &fresh_symbol = get_fresh_aux_symbol(
+    type,
+    "__goto_harness",
+    symbol_name,
+    source_locationt{},
+    initialization_config.mode,
+    goto_model.symbol_table);
+  fresh_symbol.is_lvalue = true;
+  fresh_symbol.value = init_value;
+  return fresh_symbol.symbol_expr();
+}
+
+const symbolt &recursive_initializationt::get_fresh_fun_symbol(
+  const std::string &fun_name,
+  const typet &fun_type)
+{
+  irep_idt fresh_name(fun_name);
+
+  get_new_name(fresh_name, namespacet{goto_model.symbol_table}, '_');
+
+  // create the function symbol
+  symbolt function_symbol{};
+  function_symbol.name = function_symbol.base_name =
+    function_symbol.pretty_name = fresh_name;
+
+  function_symbol.is_lvalue = true;
+  function_symbol.mode = initialization_config.mode;
+  function_symbol.type = fun_type;
+
+  auto r = goto_model.symbol_table.insert(function_symbol);
+  CHECK_RETURN(r.second);
+  return *goto_model.symbol_table.lookup(fresh_name);
+}
+
+symbolt &recursive_initializationt::get_fresh_param_symbol(
+  const std::string &symbol_name,
+  const typet &symbol_type)
+{
+  symbolt &param_symbol = get_fresh_aux_symbol(
+    symbol_type,
+    "__goto_harness",
+    symbol_name,
+    source_locationt{},
+    initialization_config.mode,
+    goto_model.symbol_table);
+  param_symbol.is_parameter = true;
+  param_symbol.is_lvalue = true;
+
+  return param_symbol;
+}
+
+symbol_exprt
+recursive_initializationt::get_symbol_expr(const irep_idt &symbol_name) const
+{
+  auto maybe_symbol = goto_model.symbol_table.lookup(symbol_name);
+  CHECK_RETURN(maybe_symbol != nullptr);
+  return maybe_symbol->symbol_expr();
+}
+
+std::string recursive_initializationt::type2id(const typet &type) const
+{
+  if(type.id() == ID_struct_tag)
+  {
+    auto st_tag = id2string(to_struct_tag_type(type).get_identifier());
+    std::replace(st_tag.begin(), st_tag.end(), '-', '_');
+    return st_tag;
+  }
+  else if(type.id() == ID_pointer)
+    return "ptr_" + type2id(type.subtype());
+  else if(type.id() == ID_array)
+  {
+    const auto array_size =
+      numeric_cast_v<std::size_t>(to_constant_expr(to_array_type(type).size()));
+    return "arr_" + type2id(type.subtype()) + "_" + std::to_string(array_size);
+  }
+  else if(type == char_type())
+    return "char";
+  else if(type.id() == ID_signedbv)
+    return "int";
+  else if(type.id() == ID_unsignedbv)
+    return "uint";
+  else
+    return "";
+}
+
+symbol_exprt recursive_initializationt::get_free_function()
+{
+  auto free_sym = goto_model.symbol_table.lookup("free");
```

Comment:
FYI we had some talk at some point about using CBMCs internal allocate/deallocate instructions for this, but I didn't have much luck for them (no need for changes, just a remark)